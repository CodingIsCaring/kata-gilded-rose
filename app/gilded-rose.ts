type Product = 'Aged Brie' | 'Sulfuras, Hand of Ragnaros' | 'Backstage passes to a TAFKAL80ETC concert';

export interface Item {
  name: Product;
  daysLeftToSell: number;
  quality: number;
}

const decreaseSellIn = (item: Item) => {
  item.daysLeftToSell = item.daysLeftToSell - 1;
};

const restartQuality = (item: Item) => {
  item.quality = 0;
};

const increaseQuality = (item: Item) => {
  item.quality = item.quality + 1;
};

const hasNotReachedTopQualityLevel = (item: Item) => item.quality < 50;

const decreaseQuality = (item: Item) => {
  item.quality = item.quality - 1;
};

const decreaseOtherProductsQuality = (item: Item) => {
  if (item.quality > 0) {
    decreaseQuality(item);
  }
  decreaseSellIn(item);
  let hasPassedSellingDate = item.daysLeftToSell < 0;
  if (hasPassedSellingDate && item.quality > 0) {
    decreaseQuality(item);
  }
};

const isTheConcertDueInLessThan = (item: Item, days: number) => item.daysLeftToSell < days
  && hasNotReachedTopQualityLevel(item);

const updateConcertTicketsQuality = (item: Item) => {
  if (hasNotReachedTopQualityLevel(item)) {
    increaseQuality(item);
    if (isTheConcertDueInLessThan(item, 11)) {
      increaseQuality(item);
    }
    if (isTheConcertDueInLessThan(item, 6)) {
      increaseQuality(item);
    }
  }
  if (item.name != 'Sulfuras, Hand of Ragnaros') {
    decreaseSellIn(item);
  }
  const hasPassedSellingDate = item.daysLeftToSell < 0;
  if (hasPassedSellingDate) {
    restartQuality(item);
  }
};

const updateAgedBrieQuality = (item: Item) => {
  if (hasNotReachedTopQualityLevel(item)) {
    increaseQuality(item);
  }
  decreaseSellIn(item);
  let hasPassedSellingDate = item.daysLeftToSell < 0;
  if (hasPassedSellingDate) {
    if (hasNotReachedTopQualityLevel(item)) {
      increaseQuality(item);
    }
  }
};

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    for (let i = 0; i < this.items.length; i++) {
      let item = this.items[i];
      let productName = item.name;
      if (productName == 'Backstage passes to a TAFKAL80ETC concert') {
        updateConcertTicketsQuality(item);
        continue;
      }

      if (productName == 'Aged Brie') {
        updateAgedBrieQuality(item);
        continue;
      }

      if (item.name != 'Sulfuras, Hand of Ragnaros') {
        decreaseOtherProductsQuality(item);
      }
    }

    return this.items;
  }
}
