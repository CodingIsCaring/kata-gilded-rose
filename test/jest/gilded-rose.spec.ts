import {GildedRose, Item} from '@/gilded-rose';

describe('Gilded Rose', () => {
  it('should do a snapshot for 2 days', function () {
    const items: Item[] = [
      {name: "Aged Brie", daysLeftToSell: 2, quality: 0},
      {name: "Sulfuras, Hand of Ragnaros", daysLeftToSell: 0, quality: 80},
      {name: "Sulfuras, Hand of Ragnaros", daysLeftToSell: -1, quality: 80},
      {name: "Backstage passes to a TAFKAL80ETC concert", daysLeftToSell: 15, quality: 20},
      {name: "Backstage passes to a TAFKAL80ETC concert", daysLeftToSell: 10, quality: 49},
      {name: "Backstage passes to a TAFKAL80ETC concert", daysLeftToSell: 5, quality: 49}];

    const gildedRose = new GildedRose(items);

    gildedRose.updateQuality();
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();
  });

  it('should do a snapshot for 10 days', function () {
    const items: Item[] = [
      {name: "Aged Brie", daysLeftToSell: 2, quality: 0},
      {name: "Sulfuras, Hand of Ragnaros", daysLeftToSell: 0, quality: 80},
      {name: "Sulfuras, Hand of Ragnaros", daysLeftToSell: -1, quality: 80},
      {name: "Backstage passes to a TAFKAL80ETC concert", daysLeftToSell: 15, quality: 20},
      {name: "Backstage passes to a TAFKAL80ETC concert", daysLeftToSell: 10, quality: 49},
      {name: "Backstage passes to a TAFKAL80ETC concert", daysLeftToSell: 5, quality: 49}];

    const gildedRose = new GildedRose(items);

    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();
  });
});
