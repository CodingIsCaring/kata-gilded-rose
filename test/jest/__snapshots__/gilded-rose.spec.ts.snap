// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose should do a snapshot for 2 days 1`] = `
Array [
  Object {
    "daysLeftToSell": 0,
    "name": "Aged Brie",
    "quality": 2,
  },
  Object {
    "daysLeftToSell": 0,
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
  },
  Object {
    "daysLeftToSell": -1,
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
  },
  Object {
    "daysLeftToSell": 13,
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 22,
  },
  Object {
    "daysLeftToSell": 8,
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
  },
  Object {
    "daysLeftToSell": 3,
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
  },
]
`;

exports[`Gilded Rose should do a snapshot for 10 days 1`] = `
Array [
  Object {
    "daysLeftToSell": -8,
    "name": "Aged Brie",
    "quality": 18,
  },
  Object {
    "daysLeftToSell": 0,
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
  },
  Object {
    "daysLeftToSell": -1,
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
  },
  Object {
    "daysLeftToSell": 5,
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 35,
  },
  Object {
    "daysLeftToSell": 0,
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
  },
  Object {
    "daysLeftToSell": -5,
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
  },
]
`;
